### ISSUE TITLE
Raise an event when a WorkloadPriorityClass.kueue.x-k8s.io is not found

### ISSUE THREAD

**Issue by akram** [kubernetes-sigs/kueue#3439]
**What would you like to be added**:
When a job is created with a specified  `kueue.x-k8s.io/priority-class` that does not exist ; then an event should be raised with a `Warning` level in the `namespace` of the `job` and with the job as the `involvedObject.name`

**Why is this needed**:
When such a job is created nothing happens and it is difficult for the user to understand what is wrong except by reading the kueue controller manager logs. As this may require admin privileges, user may just remain blocked while searching for information.

**Completion requirements**:

The keueue logs show the following error message:
```
{"level":"error","ts":"2024-11-04T18:56:42.686021656Z","caller":"controller/controller.go:329","msg":"Reconciler error","controller":"job","controllerGroup":"batch","controllerKind":"Job","Job":{"name":"low-priority-job-tg6tr","namespace":"keue-viz"},"namespace":"keue-viz","name":"low-priority-job-tg6tr","reconcileID":"1e033921-c262-401c-a462-28b74583ee51","error":"WorkloadPriorityClass.kueue.x-k8s.io \"low\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227"}

```

The relevant part being:
```
error":"WorkloadPriorityClass.kueue.x-k8s.io \"low\" not found",
```

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

**Comment by mimowo** [comment]
+1, 
I'm also wondering if we should consider deactivating such a workload as it will not get scheduled until some manual intervention. WDYT @tenzen-y ?

**Comment by mimowo** [comment]
OTOH, maybe deactivating can be considered as a breaking change for some flows when the priority class is created after the workload or in parallel. So, I'm fine for just an event now.

**Comment by tenzen-y** [comment]
> OTOH, maybe deactivating can be considered as a breaking change for some flows when the priority class is created after the workload or in parallel. So, I'm fine for just an event now.

I think that we should implement the validation webhook to verify if the specified PriorityClass exists in the cluster.
That is the same specifications as the core API PriorityClass.

**Comment by mimowo** [comment]
> I think that we should implement the validation webhook to verify if the specified PriorityClass exists in the cluster.

Technically we could but IIUC this means validation per Job CRD which is a lot of code.

> That is the same specifications as the core API PriorityClass.

I just tested it using batch/Job and referenced a non-existing `priorityClassName`, but the Pod creation failed (Kueue was not installed). Then an event is sent for the owner Job:
```
Warning  FailedCreate  4s (x5 over 19s)  job-controller  Error creating: pods "sample-job-a-qsvvf-" is forbidden: no PriorityClass with name non-existing was found
```
So, on vanilla k8s the Job creation isn't blocked, and we get an event on the Job object. I think reproducing this is reasonable.

**Comment by tenzen-y** [comment]
> So, on vanilla k8s the Job creation isn't blocked, and we get an event on the Job object. I think reproducing this is reasonable.

Yeah, I indicated the Pod creation:

```shell
cat <<EOF | k apply -f -
> apiVersion: v1
kind: Pod
metadata:
  name: high-priority-pod
  namespace: default
spec:
  priorityClassName: high-priority
  containers:
    - name: nginx
      image: nginx:latest
      ports:
        - containerPort: 80
> EOF
Error from server (Forbidden): error when creating "STDIN": pods "high-priority-pod" is forbidden: no PriorityClass with name high-priority was found
```

**Comment by mimowo** [comment]
I see, but IIUC `WorkloadPriorityClass` does not translate to Pods, so we wouldn't be able to capture that in pod webhook.

**Comment by mimowo** [comment]
Yes, we definitely want the event, the only thing is that if we deactivate then we already have a mechanism to propagate the event to the Job object, so the solution will be shorter by reusing this path, and the information will be reflected both in event and status. wdyt @tenzen-y ?

**Comment by mimowo** [comment]
sorry, I think this is not directly related to dashboard
/kind-remove dashboard

**Comment by tenzen-y** [comment]
> Yes, we definitely want the event, the only thing is that if we deactivate then we already have a mechanism to propagate the event to the Job object, so the solution will be shorter by reusing this path, and the information will be reflected both in event and status. wdyt [@tenzen-y](https://github.com/tenzen-y) ?

Deactivating the corresponding Workload sounds reasonable. 

Batch user PoV, if we do not deactivate the Workload, and it is never admitted, it's challenging to identify the reason why their Workload is never admitted.

Cluster Admin PoV, they do not want to handle the Workload with non-existence WoriloadPriorityClass as default Priority (1) since it potentially could break priority design in the cluster.

**Comment by tenzen-y** [comment]
This looks like a still valid request.
/remove-lifecycle stale

**Comment by tenzen-y** [comment]
I'm not sure if this is a good first issue since the contributor must implement the Workload deactivation mechanism.

**Comment by j4ckstraw** [comment]
I'd like to do some work.

**Comment by tenzen-y** [comment]
> I'd like to do some work.

@j4ckstraw Thank you for your interest in this issue. Feel free to take this one with `/assign` comment.

**Comment by mimowo** [comment]
/reopen
https://github.com/kubernetes-sigs/kueue/pull/13645 didn't intend to close the issue

**Comment by thc1006** [comment]
Two PRs point at this issue now, so a short map in case it saves anyone time.

#6688 is @j4ckstraw's original and covers the event together with admission-time rejection. #13775 is the event-only split @mbobrovskyi asked for, written against current `main`, with unit and integration coverage.

Neither proposes anything new here. @mimowo landed on wanting the event in this thread back in 2024, and @akram's point that admission cannot protect against a `WorkloadPriorityClass` being deleted afterwards is why it reports rather than rejects.

@j4ckstraw, I have not asked for the assignment and am not asking now. If you would rather carry the event part yourself, say so and I will close mine.

Written with AI assistance.

**Comment by wryyyds7** [comment]
I'm working on this. The plan is to emit a Warning Event on the Job object when the referenced WorkloadPriorityClass is not found, so users can diagnose the issue via `kubectl get events` without needing access to controller logs.

The change targets two call sites in `pkg/controller/jobframework/reconciler.go` where `prepareWorkload` returns a NotFound error: `handleJobWithNoWorkload` (workload creation) and `updateWorkloadToMatchJob` (workload update). Both will now emit a `PriorityClassNotFound` Warning Event before returning the error.
